kernel/src/main/java/org/kframework/compile/GatherVarsVisitor.java

```diff
@@ -37,6 +41,14 @@ public void apply(KApply k) {
         if (k.klabel() instanceof KVariable) {
             apply((KVariable) k.klabel());
         }
+        if (k.klabel().equals(KLabels.ML_EXISTS) || k.klabel().equals(KLabels.ML_FORALL)) {
+            boolean tmp = isInExistsLhs;
+            isInExistsLhs = true;
+            apply(k.items().get(0));
```

Comment:
Why are we triggering re-traversal of the variable part of the pattern? I would think that we could just get it, assert that it's a variable (indeed, it's not a well-formed pattern if it's not), and then extract it directly. No need to recurse on that subterm.

I think that recursing on th esubterm is also why you have to have the `tmp` variable for `isInExistsLhs`, which makes this more confusing.

Reply:
It also makes it more confusing because then `KVariable` case needs to be adjusted. I think you can do without adjusting that case if you change the code to say something like this (psuedocode):

```suggestion
    binderVar = k.items().get(0);
    if type(binderVar) is not KVariable {
        errors.add(KEMException.compilerError("Found non-variable in binder position of {LABEL}");
    }
    vars.add(binderVar);
```

Then we don't need the variable `isInExistsLhs`, and don't need to modify `KVariable` case.

Reply:
You can have casts and injections. And maybe brackets.
It's a similar pattern we use in [ComputeUnboundVariables](https://github.com/runtimeverification/k/blob/master/kernel/src/main/java/org/kframework/compile/checks/ComputeUnboundVariables.java#L69)
If you write something like `  rule foo(X) => #Exists 1 . { ?I #Equals 1 }` then the haskell-backend complains.
```
kore-exec: [16411] Error (ErrorParse):
    definition.kore:725:133: unexpected token TokenDomainValue
```
Not the pretties error message, but kind of gets out of scope for this issue.

Reply:
I don't think it's out of scope for this issue at all, this issue is about handling Exists and Forall binding correctly.

I also just think that the implementation I propose above is cleaner honestly, and easier to understand. It took me a chunk of time to understand your code, with the temp variable being stored and unset. I don't think the check is actually complicated, but the way you implemented it made it seem so.

Reply:
Is it actually reasonable to have a cast/injection/bracket on the first argument here? To me it seems like not at first glance. But perhaps it makes sense?

Reply:
> I don't think it's out of scope for this issue at all, this issue is about handling Exists and Forall binding correctly.

Fair point.
```
rule: rule 1 => #Exists _Y . 2
parsed: rule #token("1","Int")=>#Exists(#SemanticCastToK(_Y),#token("2","Int")) requires #token("true","Bool") ensures #token("true","Bool") [org.kframework.attributes.Location(Location(18,8,18,27)), org.kframework.attributes.Source(Source(/home/radu/work/test/test.k)), org.kframework.definition.Production(syntax #RuleContent ::= #RuleBody [klabel(#ruleNoConditions), symbol])]
```
Casts will appear under there. They are added by the type inferencer even if you don't provide one.
I'm not sure about injections since that's a free parameter so the type inferencer can choose something convenient.
Brackets probably not since we remove them right after parsing.

But in any case, this type of traversal has proven to be the most flexible and reliable for the front-end. If we decide in the future that we want to support a list of variables `#Exists A, B, C . 2` then this will still work correctly.

Anyway, thanks for the approval.